### PR TITLE
Add clip, transform, save/restore methods

### PR DESCRIPTION
**Pull request by raphlinus**
This patch adds several fundamental methods to RenderContext and implementations to the three back-ends: clip, save/restore, and transform.

### CODE REVIEW

File: piet-cairo/src/lib.rs [hunk 1]
```diff
@@ -165,6 +171,20 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
...
+
+    fn save(&mut self) {
+        self.ctx.save();
+    }
+
+    fn restore(&mut self) {
+        self.ctx.restore();
+    }
+
+    fn finish(&mut self) {}
```
Comment (rylev) [inline piet-cairo/src/lib.rs]: Will the cairo and web versions of finish always be empty? I worry about learnability and useability when forgetting to call `finish` on web and cairo are inconsequential but could break things on direct-2d

File: piet-direct2d/examples/basic.rs [hunk 2]
```diff
@@ -51,6 +68,16 @@ fn draw_pretty_picture<R: RenderContext>(rc: &mut R) {
         1.0,
         None,
     );
+
+    rc.save();
```
[the remark this answers is not in the document]
Comment (rylev) [reply piet-direct2d/examples/basic.rs]: I think it's ok to have both APIs (where `save`/`restore` should be implemented by backends and the convenience method is implemented in terms of them). You could mention in docs that the convenience function is prefered but if people are used to the PostScript-like API they're free to use it.

File: piet-direct2d/src/lib.rs [hunk 3]
```diff
@@ -50,6 +63,18 @@ impl<'a> D2DRenderContext<'a> {
...
+            ctx_stack: vec![CtxState::default()],
+        }
+    }
+
+    fn current_transform(&self) -> Affine {
+        self.ctx_stack.last().unwrap().transform
+    }
+
+    fn pop_state(&mut self) {
+        let old_state = self.ctx_stack.pop().unwrap();
```
[the remark this answers is not in the document]
Comment (rylev) [reply piet-direct2d/src/lib.rs]: I filed this: https://github.com/linebender/piet/issues/20 Feel free to close to add your thoughts or close it in favor of something more descriptive. 😄

**Comment by raphlinus** [comment]
#20 looks like a great place to continue the error discussion. Now that's captured and I added `with_save`, think this is good to go?

**Comment by rylev** [comment]
Looking great!

File: piet-cairo/src/lib.rs (hunk 1 above)
Comment (SimonSapin) [reply piet-cairo/src/lib.rs]: > a utility function that finishes the render context right before extracting the bitmap

This sounds like https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-flush, but there’s also something named `cairo_surface_finish`.

**Pull request merged**